File: 2024/02/day2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <fstream>

#define INPUT_PATH "inputs/2.txt"

static void readFileToVector(const char* filename, std::vector<std::vector<int>>& matrix) {
    for (auto vector : matrix) {
        vector.clear();
    }

    matrix.clear();

    std::ifstream file(filename);
    std::string line;

    while (getline(file, line)) {
        std::stringstream numStringStream(line);
        std::string numString;
        std::vector<int> numsVector;

        while (getline(numStringStream, numString, ' ')) {
            numsVector.push_back(stoi(numString));
        }

        matrix.push_back(numsVector);
    }
}

int isVectorSafe(std::vector<int> vector) {
    int lastNum = vector[0];
    int lastDirection = (vector[0] - vector[1]) > 0;
    int safe = 1;

    for (int i = 1; i < vector.size(); ++i) {
        int direction = (lastNum - vector[i]) > 0;
        int difference = abs(lastNum - vector[i]);

        if (difference == 0 || difference > 3 || direction != lastDirection) {
            safe = 0;
            break;
        }

        lastNum = vector[i];
    }

    return safe;
}

int solvePart1(std::vector<std::vector<int>>& matrix) {
    int safeReports = 0;

    for (auto vector : matrix) {
        safeReports += isVectorSafe(vector);
    }

    return safeReports;
}

int solvePart2(std::vector<std::vector<int>>& matrix) {
    int safeReports = 0;

    for (auto vector : matrix) {
        for (int i = 0; i < vector.size(); ++i) {
            std::vector<int> possibleSubvector;

            for (int j = 0; j < vector.size(); ++j) {
                if (j == i) continue;
                possibleSubvector.push_back(vector[j]);
            }
            
            if (isVectorSafe(possibleSubvector)) {
                safeReports++;
                break;
            }
        }
    }

    return safeReports;
}

int main() {
    std::vector<std::vector<int>> matrix;
    readFileToVector(INPUT_PATH, matrix);

    int part1Solution = solvePart1(matrix);

    std::cout << part1Solution << std::endl;

    int part2Solution = solvePart2(matrix);

    std::cout << part2Solution << std::endl;

    return 0;
}

File: 2024/02/day2.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

const INPUT_PATH: &str = "inputs/2.txt";
fn main() {
    let matrix = lines_to_i8_matrix(INPUT_PATH);

    let part_1_solution = solve_part_1(&matrix);
    println!("{}", part_1_solution);

    let part_2_solution = solve_part_2(&matrix);
    println!("{}", part_2_solution);
}

fn solve_part_1(matrix: &Vec<Vec<i8>>) -> i32 {
    let mut safe_reports = 0;
    for vec in matrix {
        if is_vector_safe(vec) {
            safe_reports += 1;
        }
    }

    safe_reports
}

fn solve_part_2(matrix: &Vec<Vec<i8>>) -> i32 {
    let mut safe_reports = 0;

    for vec in matrix {
        for i in 0..vec.len() {
            let mut possible_subvector: Vec<i8> = vec![];

            for j in 0..vec.len() {
                if j == i {
                    continue;
                }
                possible_subvector.push(vec[j]);
            }

            if is_vector_safe(&possible_subvector) {
                safe_reports += 1;
                break;
            }
        }
    }

    safe_reports
}

fn is_vector_safe(vec: &Vec<i8>) -> bool {
    let mut last_num = vec[0];
    let initial_direction = (vec[0] - vec[1]) > 0;

    for num in &vec[1..] {
        let direction = (last_num - *num) > 0;
        let value: i8 = last_num - *num;
        let difference = value.abs();

        let is_unsafe = difference == 0 || difference > 3 || direction != initial_direction;

        if is_unsafe {
            return false;
        }

        last_num = *num;
    }

    true
}

fn lines_to_i8_matrix(path: &str) -> Vec<Vec<i8>> {
    if let Ok(lines) = read_lines(path) {
        let mut matrix: Vec<Vec<i8>> = vec![];
        for line in lines.flatten() {
            let mut vec = Vec::new();
            for num in line.split_whitespace() {
                let num: i8 = num
                    .trim()
                    .parse()
                    .expect("Invalid number, something is wrong with the input!");
                vec.push(num);
            }
            matrix.push(vec);
        }
        return matrix;
    }
    vec![vec![]]
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}
